src/finance_quote/modules/fixerio.py

```diff
@@ -0,0 +1,185 @@
+except ImportError: import json
+
+
+class FixerioQuote(Quote):
+    """ Fixer.io-specific quote """
+    pass
+
+
+class Fixerio(Source):
+    """Retrieves prices from data files or online provider(s)"""
+    def __init__(self):
+        self.cache_path = tempfile.gettempdir()
+        self.logger = logging.getLogger(__name__)
+
+    def download(self, namespace: str, mnemonic: str, currency: str) -> FixerioQuote:
+        """ Download latest rates. Caches into temp directory. """
+        namespace = namespace.upper()
+        assert namespace == "CURRENCY"
+        # namespace is ignored, anyways.
+        currency = currency.upper()
+        mnemonic = mnemonic.upper()
+        # make sure the symbol does not contain namespace
+        if ":" in mnemonic:
+            raise ValueError("Currency symbol should not contain namespace.")
+
+        rates_dict = None
+        if self.latest_rates_exist():
+            self.logger.debug("Cached rates found.")
+            rates_dict = self.__read_rates_from_file()
+        else:
+            rates_dict = self.__download_rates(currency)
+
+        mapper = FixerioModelMapper(rates_dict)
+        model = mapper.get_model_for_symbol(mnemonic)
+        self.logger.debug(model)
+        return model
+
+    def __download_rates(self, base_currency: str):
+        """
+        Downloads the latest rates. Requires base currency and a list of currencies to
+        retrieve.
+        Returns json response object from Fixer.io.
+        """
+        import requests
+
+        assert isinstance(base_currency, str)
+
+        # Downloads the latest rates using Fixerio. Returns dict.
+        # https://pypi.python.org/pypi/fixerio
+        # from fixerio import Fixerio
+        # fxrio = Fixerio(base=base_currency) #, symbols=symbols)
+        # latest_rates = fxrio.latest()
+
+        result = None
+        base_url = 'http://api.fixer.io/latest'
+        query = base_url + "?base=" + base_currency
+
+        # if symbols:
+        #     symbols_csv = ",".join(symbols)
+        # if symbols_csv:
+        #     query += f"&symbols={symbols_csv}"
+
+        try:
+            self.logger.debug("retrieving rates from %s" % query)
+            response = requests.get(query)
+            # print("[%s] %s" % (response.status_code, response.url))
+            if response.status_code != 200:
+                result = 'N/A'
+            else:
+                result = response.json()
+                # rate_in_currency = rates["rates"][rate_in]
+        except requests.ConnectionError as error:
+            self.logger.error(error)
+
+        self.logger.debug("Latest prices downloaded.")
+
+        # Since these are daily rates, cache them into a file.
+        # Ignored for now since we are downloading individual currencies, not all together.
+        self.__save_rates(result)
+
+        return result
+
+    def get_yesterdays_file_path(self):
+        """ Full path to the today's rates file. """
```

Comment:
all the caching features add some complexity.
it would be better IMO to have a light fixerio model that has not caching whatsoever and then have another configurable class implementing the caching that could be reuse for other modules too.

Reply:
That's true. Do you have any specific concerns?

Reply:
You mean concrete examples of concerns? If so, unit testing will be difficult. Disabling caching (because fixerio has fixed a buggy exchange rate but there is caching, difficult to call again the original source), need file creation permission (maybe the filesystem where the lib is installed is in read-only)

Reply:
All true. However, I have no intention of pre-fixing somebody else's bugs nor permissions. If somebody runs into that situation, I'd happily accept a pull request with the fix that is backwards-compatible.